rfd/0180-sso-mfa.md

```diff
@@ -0,0 +1,833 @@
+enforcement itself. Continue? (y/N):
+```
+
+Additionally, forced re-authentication will be the default setting. Even if a
+connector is enabled for MFA without actual strict IdP MFA checks, the user will
+still be required to re-authenticate through the SSO provider. In practice,
+this means that a stolen SSO session is not enough for an attacker to bypass
+MFA checks, they'd need to steal the user's SSO identity (password, MFA).
+
+#### IdP Compromise
+
+In the case of a full-scale IdP compromise, an attacker may have the ability
+to auto-provision users with arbitrary permissions.
+
+When device trust is required, newly auto-provisioned SSO users are required
+to add their first MFA device from a trusted device. When combined with MFA
+security features, such as MFA for Admin actions and per-session MFA, the blast
+radius of an IdP compromise is largely contained. The attacker would be
+prevented from accessing any critical infrastructure or making any changes
+to the cluster's security configuration.
+
+Allowing SSO as an MFA method would bypass the device trust check, opening the
+cluster back up to attacks in the case of an IdP compromise. To maintain this
+invariant, device trust must be enforced within the SSO MFA check.
+
+### Implementation
+
+#### SSO MFA flow
+
+```mermaid
+sequenceDiagram
+    participant Client
+    participant Teleport
+    participant IdP
+    participant Node
+
+    Note over Client: `tsh ssh node`
+    Note over Client: Prepare SSO client callback URL
+    par initiate SSO MFA ceremony
+        Client ->> Teleport: rpc CreateAuthenticateRequest<br/>w/ client callback url
+        Teleport ->> Teleport: create SSO Auth request
+        Teleport ->> Teleport: SSO MFA session for { request_id, username }
+        Teleport ->> Client: SSOChallenge{ redirect_url, request_id }
+    end    
+    Note over Client: Prompt user to complete SSO MFA check
+    par complete SSO MFA ceremony
+        Client ->> IdP: browse to redirect URL
+        IdP ->> IdP: login and pass configured MFA check
+        IdP ->> Teleport: send successful IdP auth response
+        Teleport ->> Teleport: verify IdP username matches the SSO Auth request username
+        Teleport ->> Teleport: retrieve SSO MFA session for { request_id, username }<br/>and add an mfa_token
+        Teleport ->> IdP: return client callback URL<br/>w/ encode encrypted mfa_token in query params
+        IdP -> Client: redirect
+    end
+    Note over Client: Decode and decrypt mfa_token<br/>from redirect query params
+    par per-session MFA ceremony
+        Client ->> Teleport: rpc GenerateUserSingleUseCerts<br/>w/ SSOChallengeResponse{ request_id, mfa_token }
+        Teleport ->> Teleport: retrieve SSO MFA session for { request_id, username }
+        Teleport ->> Teleport: check mfa_token against SSO MFA session
+        Teleport ->> Client: issue MFA verified certs
+    end
+    Client ->> Node: connect w/ MFA verified certs
+```
+
```

Comment:
Not clear how the successful IdP Auth response should go directly to Teleport. Same for the callback at step 9.

Additionally:
- since the SSO MFA flow and the per-session MFA flow are separated, it is not clear how the final step is protected against phishing. Probably it is in the plan by using the request ID, but I would state it as requirement in the RFD
- How will replay attempts be treated? As per-session, the token should not be usable two times in a row, same for the certs. What is the plan in that sense (I would add it to the RFD)

Reply:
> since the SSO MFA flow and the per-session MFA flow are separated, it is not clear how the final step is protected against phishing. Probably it is in the plan by using the request ID, but I would state it as requirement in the RFD

For `tsh` and Teleport Connect, the client passes a secret key in the login request, and the server encodes the MFA token with that key before sending it back. The request id is also provided by the server and should only be known by the requesting client, the mfa token alone is useless without knowledge of the request id.

For the Web, we don't encode the mfa token with a secret key, we just depend on the request id being unknown by anyone but the client.

Also note that the mfa token is only valid for the user who requested it. So a user with active key+certs can cash in the mfa token w/ request ID to get their MFA verified certs.

Does this all sound good? This was my main concern that I wanted a second opinion on, so thanks for spotting it out. I'll update the document with agreed upon details.

> How will replay attempts be treated? As per-session, the token should not be usable two times in a row, same for the certs. What is the plan in that sense (I would add it to the RFD)

For the most part, the MFA token will be discarded as soon as it's verified for mfa-verified certs, etc. However, we also track the [MFA scope](https://github.com/gravitational/teleport/blob/master/rfd/0155-scoped-webauthn-credentials.md) for SSO MFA requests and [allow reuse](https://github.com/gravitational/teleport/blob/master/rfd/0155-scoped-webauthn-credentials.md#reuse) in certain contexts. e.g. MFA for back-to-back admin actions: create user -> create reset password token for registration.

Let me know if you are not convinced that SSO MFA is phishing resistant enough to allow reuse like we do with WebAuthn, with the details above in mind.